### PR TITLE
FIRMessagingRemoteNotificationsProxy - test only public API with no assumptions on implementation

### PR DESCRIPTION
**Pull request by maksymmalyhin**
The update is an intermediate step for #2591. To update `FIRMessagingRemoteNotificationsProxy` implementation with `GoogleUtilities` we need reliable set of tests.

- `FIRMessagingRemoteNotificationsProxyTests` updated to use only public API of `FIRMessagingRemoteNotificationsProxy`. This will allow to reuse the same tests for the updated implementation.

### CODE REVIEW

File: Example/Messaging/Tests/FIRMessagingRemoteNotificationsProxyTest.m [hunk 1]
```diff
@@ -24,38 +24,30 @@
...
-void FCM_swizzle_appDidReceiveRemoteNotificationWithHandler(
-    id self, SEL _cmd, UIApplication *app, NSDictionary *userInfo,
-    void (^handler)(UIBackgroundFetchResult));
-void FCM_swizzle_willPresentNotificationWithHandler(
-    id self, SEL _cmd, id center, id notification, void (^handler)(NSUInteger));
-void FCM_swizzle_didReceiveNotificationResponseWithHandler(
-    id self, SEL _cmd, id center, id response, void (^handler)());
+#pragma mark - Invalid App Delegate or UNNotificationCenter
 
+@interface RandomObject : NSObject
```
Comment (chliangGoogle) [inline Example/Messaging/Tests/FIRMessagingRemoteNotificationsProxyTest.m]: Not sure why we need to define a class for this, the purpose of passing a randomObject is to passing an invalid class, defining a serials of methods seem unnecessary.

File: Example/Messaging/Tests/FIRMessagingRemoteNotificationsProxyTest.m [hunk 2]
```diff
@@ -24,38 +24,30 @@
 #import "FIRMessaging.h"
 #import "FIRMessagingRemoteNotificationsProxy.h"
 
-#pragma mark - Expose Internal Methods for Testing
-// Expose some internal properties and methods here, in order to test
-@interface FIRMessagingRemoteNotificationsProxy ()
```
Comment (chliangGoogle) [inline Example/Messaging/Tests/FIRMessagingRemoteNotificationsProxyTest.m]: why remove all these? I think this is a standard way if you want to access private methods for testing.

File: Firebase/Messaging/FIRMessaging.m [hunk 3]
```diff
@@ -259,7 +259,7 @@ - (void)configureMessaging:(FIRApp *)app {
                              @"proper integration.",
                              kFIRMessagingRemoteNotificationsProxyEnabledInfoPlistKey,
                              docsURLString);
-    [FIRMessagingRemoteNotificationsProxy swizzleMethods];
+    [[FIRMessagingRemoteNotificationsProxy sharedProxy] swizzleMethodsIfPossible];
```
Comment (chliangGoogle) [inline Firebase/Messaging/FIRMessaging.m]: why? A class method seems simpler.

File: Example/Messaging/Tests/FIRMessagingRemoteNotificationsProxyTest.m [hunk 4]
```diff
@@ -154,30 +154,36 @@ - (void)tearDown {
 #pragma mark - Method Swizzling Tests
 
 - (void)testSwizzlingNonAppDelegate {
-  id randomObject = @"Random Object that is not an App Delegate";
-  [self.proxy swizzleAppDelegateMethods:randomObject];
-  XCTAssertFalse(self.proxy.didSwizzleAppDelegateMethods);
-}
+  RandomObject *invalidAppDelegate = [[RandomObject alloc] init];
```
Comment (chliangGoogle) [inline Example/Messaging/Tests/FIRMessagingRemoteNotificationsProxyTest.m]: Your solution for each tests work but I'm not sure what's wrong with the original one?
We need a good reason of refactoring existing tests, not just for the sake of refactor. It takes time for writing, reviewing and maintaining the release for big changes.

**Comment by chliangGoogle** [comment]
I'm not following how these changes related to the requirement in  #2591.

Looks like you are just refactoring the testing code but I'm not 100% clear on what you are trying to do here. Can you elaborate how these big refactors would help with  #2591?
We should not refactor just for the sake of refactoring, maybe I'm not seeing the reason under the hood.

**Comment by maksymmalyhin** [comment]
@chliangGoogle Thank you for the review.

As I stated in the PR description, I updated tests to avoid relaying on private methods and variable which are going to change significantly as result of #2591. 

I posted this PR first to separate changes and simplify the review process. So, in this branch we have old implementation proofed by the tests. The next stage will be updating the implementation without changing the tests. If tests rely on the implementation of the class, then the tests cannot be reused to verify the implementation update (which I actually encountered).

Please let me know if you still have concerns.

File: Example/Messaging/Tests/FIRMessagingRemoteNotificationsProxyTest.m (hunk 4 above)
Comment (chliangGoogle) [reply Example/Messaging/Tests/FIRMessagingRemoteNotificationsProxyTest.m]: Talked offline and looks like we only want to use public API for testing to be able to switch the new swizzling class. Will close some of the conversations.

File: Firebase/Messaging/FIRMessagingRemoteNotificationsProxy.h [hunk 5]
```diff
@@ -31,10 +31,15 @@
...
+/**
+ * A shared instance of `FIRMessagingRemoteNotificationsProxy`
+ */
++ (instancetype)sharedProxy;
+
 /**
  *  Swizzles Application Delegate's remote-notification callbacks and User Notification Center
  *  delegate callback, and invokes the original selectors once done.
  */
-+ (void)swizzleMethods;
```
Comment (chliangGoogle) [inline Firebase/Messaging/FIRMessagingRemoteNotificationsProxy.h]: Can we also remove the implementation of swizzleMethods?

File: Example/Messaging/Tests/FIRMessagingRemoteNotificationsProxyTest.m (hunk 1 above)
Comment (maksymmalyhin) [reply Example/Messaging/Tests/FIRMessagingRemoteNotificationsProxyTest.m]: @chliangGoogle  To verify if the object methods were swizzled without using private details we need to invoke one of the methods and check its side effect.
E.g. on the [line 163](https://github.com/firebase/firebase-ios-sdk/blob/ad1da54d0fb34f72fb3139ef833a711fb8d555b1/Example/Messaging/Tests/FIRMessagingRemoteNotificationsProxyTest.m#L163) we perform the method and verify if [ [[FIRMessaging messaging] appDidReceiveMessage:userInfo] ](https://github.com/firebase/firebase-ios-sdk/blob/ad1da54d0fb34f72fb3139ef833a711fb8d555b1/Example/Messaging/Tests/FIRMessagingRemoteNotificationsProxyTest.m#L161) was invoked. If it was, that would mean that even the random object is swizzled.

Defining the methods on an object which does not conform to the required protocols looks like the simples way to implement it. Please let me know if you can see an easier way.

**Comment by maksymmalyhin** [comment]
@chliangGoogle Thank you for the attentive review!

**Pull request merged**